### DISCUSSION TITLE
First time using MonoAlg3D

### DISCUSSION THREAD

**Discussion by edorachlee** [rsachetto/MonoAlg3D_C#72] [General]
Hi everyone,

I'm a PhD student from the Silva Lab at Washington University in St. Louis. While I do have experience with EP simulations, this is my first time using MonoAlg3D. Is there a basic documentation of the available parameters in the .ini files, i.e what they are and possible values for these parameters? I am also quite lost on how I should know all the different types of functions out there("update_monodomain_default", "init_conjugate_gradient", etc.). Perhaps the documentation is there and I just can't find it because I'm dumb. Any help (and your understanding for a newbie) is much appreciated! Thank you.

Best,
Lucas Lee

**Comment by rsachetto** [comment]
Dear Lucas, we lack good software documentation and you are not dumb :).
I've attached a presentation I made with some information about the parameters. 

[presentation.pdf](https://github.com/user-attachments/files/16785914/presentation.pdf)

Please let me know if I can help you with anything else.
Best,


Rafael

**Comment by edorachlee** [comment]
Thanks Rafael! I've gone through your presentation and it helped me, thank you.

I have been trying to load a .vtk mesh for simulations. From your presentation it looks like the mesh converter is able to support VTK -> ALG conversion. However, when I tried this with the rabbit heart mesh, I am only able to do ALG -> VTK and not the other way. Could I confirm this is the expected behavior?

**Comment by rsachetto** [comment]
The simulator is extensible, so creating a function to load VTK meshes would be possible. I decided not to ship a function like this, to avoid having the VTK lib as a software dependency, so you have to convert the VTK mesh first. 

We have documentation about this process:

https://github.com/rsachetto/MonoAlg3D_C/wiki/Loading-a-VTU-mesh

Let me know if I can help with anything else.

My best,

Rafael

**Reply by edorachlee** [reply]
This is very helpful, thanks.

**Reply by edorachlee** [reply]
Hi Rafael, I have been following along the custom mesh loading process. So far I have successfully converted some .vtk files containing fiber information into .vtu and then .alg files.

When using the SET_SPATIAL_DOMAIN function to load the mesh, it seems that we need to hardcode the number of volumes in the custom mesh. Is there a way to dynamically pass the number of volumes? With a large number of custom meshes, I believe this may be a problem in the future...

Best,
Lucas

**Reply by rsachetto** [reply]
Hi Lucas. Are you developing your own function or using initialize_grid_with_custom_mesh? 
In both cases, You can pass the number of volumes in the .ini file using the number_of_points parameter.

**Reply by edorachlee** [reply]
Ah, I see. If I generate unique .ini files for each mesh, I can avoid this problem by defining number_of_points in each file. Thank you. (I was writing my custom function)

**Reply by edorachlee** [reply]
As a follow-up, if I have custom data in the .vtk files (such as fiber orientation), will initialize_grid_with_custom mesh still be able to accept the additional information? Assuming that I've created a custom struct object and corresponding SET_CUSTOM_DATA_FOR_MESH function.

**Reply by rsachetto** [reply]
If you are using only the fiber orientation, you do not need a custom struct, as this information is located in the cell, if I am not mistaken. You have to create a custom function to add this data in each cell.

**Reply by edorachlee** [reply]
Ah, I also have one additional field (element tag) that defines whether the cell is the LV or the RV. I could get rid of the information but I am trying to keep it for now...

If I have to make a custom struct, then will I need to make my custom function instead of using initialize_grid_with_custom_mesh?

**Reply by rsachetto** [reply]
You can use initialize_grid_with_custom_mesh. But you have to pass a function implemented with SET_CUSTOM_DATA_FOR_MESH macro to the initialize_grid_with_custom_mesh. This function will be called for each cell and initialize the cell->mesh_extra_info data. This data will point to your custom struct. A I away from the computer now, but latter I can send you some examples.

**Reply by rsachetto** [reply]
This an example of a SET_CUSTOM_DATA_FOR_MESH

```
SET_CUSTOM_DATA_FOR_MESH(set_custom_data) {

     //Here the index of the custom_data array is the position of the data after the 
     //mesh geometrical information. For instance,  if the first column after the geometrical data
     //is the element tag, will would access it using custom_data[0];

    cell->sigma.fibers.f[0] = custom_data[0];
    cell->sigma.fibers.f[1] = custom_data[1];
    cell->sigma.fibers.f[2] = custom_data[2];
}
```

Please, let me know if I can help with anything else. If you need we can schedule a call so that I can help you to solve your
doubts.

**Reply by edorachlee** [reply]
Ok, I understand now. In conjunction with the documentation I believe I should be good for now. Thank you.

**Reply by edorachlee** [reply]
Continuing our discussion on the fibers...after converting a mesh with fiber information, is there a good way to visualize the fiber orientations in the .ALG file?

**Reply by rsachetto** [reply]
Unfortunately it is not possible to draw the fiber orientation in the visulalizer yet. You can only see the value of each coordinate, by pressing the numbers 0 to 9 in the keyboard. This feature is in the roadmap.

**Comment by edorachlee** [comment]
Hi Rafael, I wanted to jump back on this thread to discuss the visualizer. Sometimes, when I load in my simulation results, I can't see the mesh. However, I can still click on the Play button and can see that the time is still being incremented. Just can't see the mesh. This happens frequently but not all the time. Any idea what's going on?

Also, is there a way to view just the .alg mesh without having to conduct a simulation? Currently I'm running into a similar situation where I don't see the mesh when I load it, and I must load some simulations together with the mesh to view it properly.

**Reply by rsachetto** [reply]
> Hi Rafael, I wanted to jump back on this thread to discuss the visualizer. Sometimes, when I load in my simulation results, I can't see the mesh. However, I can still click on the Play button and can see that the time is still being incremented. Just can't see the mesh. This happens frequently but not all the time. Any idea what's going on?

Could you send me the output files of the simulation? Sometimes, when you cancel a simulation, the output files get corrupted.

> 
> Also, is there a way to view just the .alg mesh without having to conduct a simulation? Currently I'm running into a similar situation where I don't see the mesh when I load it, and I must load some simulations together with the mesh to view it properly.

Yes, you can. Just pass the .alg file as a parameter in the command line.

**Reply by edorachlee** [reply]
Thank you very much for taking a look. Here is an example directory: 
[cuboid_ohara_gpu.zip](https://github.com/user-attachments/files/17101742/cuboid_ohara_gpu.zip)

And here is a video of me loading in the example: 
[Screencast from 2024-09-23 12-45-48.webm](https://github.com/user-attachments/assets/bdf02d06-94be-4333-aedc-c34799561b05)

This is simply the first OHara example you provide in your build instructions, but pretty much all of my experiments have the same issue. 

Unfortunately, when loading in the .alg file as a parameter, the same issue persists.

**Reply by rsachetto** [reply]
> Thank you very much for taking a look. Here is an example directory: [cuboid_ohara_gpu.zip](https://github.com/user-attachments/files/17101742/cuboid_ohara_gpu.zip)
> 
> And here is a video of me loading in the example: [Screencast from 2024-09-23 12-45-48.webm](https://github.com/user-attachments/assets/bdf02d06-94be-4333-aedc-c34799561b05)
> 
> This is simply the first OHara example you provide in your build instructions, but pretty much all of my experiments have the same issue.
> 
> Unfortunately, when loading in the .alg file as a parameter, the same issue persists.

That is strange. I just tested it here and it is working fine. What GPU are you using to run the visualizer? Maybe it is an opengl problem. I had some issues with Intel GPUs in the past.

Just to let you know. For this example, you can view the simulation using Paraview.

**Reply by edorachlee** [reply]
I have two NVIDIA 1080's in my personal computer. The OS os Ubuntu 24.04.

Thanks to your suggestion I was able to use Paraview to view my simulations. So it seems it is not a corruption problem?

Edit 1: as a reference here is my OpenGL version: 4.6.0 NVIDIA 550.107.02
Edit 2: Whenever I launch the dialog to open a file I receive this message: "Gtk-Message: 13:48:25.898: GtkDialog mapped without a transient parent. This is discouraged." Maybe this has some connection with the issue?

**Reply by rsachetto** [reply]
> I have two NVIDIA 1080's in my personal computer. The OS os Ubuntu 24.04.

I'm also using a 1080 in my PC. My os is different, but it should work in Ubuntu.


> Thanks to your suggestion I was able to use Paraview to view my simulations. So it seems it is not a corruption problem?

No, it is not a corruption problem.


> Edit 1: as a reference here is my OpenGL version: 4.6.0 NVIDIA 550.107.02 

Same version here: **OpenGL version: 4.6.0 NVIDIA 550.107.02.**

> Edit 2: Whenever I launch the dialog to open a file I receive this message: "Gtk-Message: 13:48:25.898: GtkDialog mapped without a transient parent. This is discouraged." Maybe this has some connection with the issue?

No, this is not a problem. I have the same message here.

Are you using X or wayland?

**Reply by edorachlee** [reply]
X11!

**Reply by rsachetto** [reply]
That is really strange. I've just made a clean build here and everything is working properly. What is your gcc version?

**Reply by edorachlee** [reply]
Version:
gcc (Ubuntu 13.2.0-23ubuntu4) 13.2.0

**Reply by rsachetto** [reply]
I don't think this is the problem. I'm really in the dark here. There is nothing wrong with you setup or distribution.

**Reply by edorachlee** [reply]
It's an interesting phenomenon. I can even click randomly on the screen and somehow select some vertices on the mesh, then record an AP also. 

I think this is not a major problem so we can pause here for now, as the Paraview solution currently works for me. I will continue to tinker with the software and make an update when/if I find a workaround. Thanks for your time and assistance.

**Reply by rsachetto** [reply]
There is a paraview plugin to open alg meshes using paraview. It is in scripts/paraview-alg-plugin.py . You can load it in Paraview by clicking in tools->manage plugins->load new and choose the scripts/paraview-alg-plugin.py file. You can also mark it as Auto Load.

**Reply by edorachlee** [reply]
Ahh that is also very beneficial to know and will help me very much. Thank you.

**Reply by rsachetto** [reply]
Hi Lucas, I just discovered what is happening. You are probably running the visualizer outside the root dir. I don't know why the mesh is not being shown when running the software outside the root dir.

You should run from the terminal, inside the MonoAlg3D_C folder as:

```sh
./bin/MonoAlg3D_visualizer your_mesh
```
The mesh can be located in any folder. I think the problem is that the shaders are not being loaded properly. I will try to embed them in the executable to avoid such problems.

**Reply by edorachlee** [reply]
Yes this is it! Thank you for tracking down the problem.

**Comment by benjamitchell** [comment]
Hi Rafael! how are you?
My name is Benja mitchell, i'm a civil engineering mathematics student. I wanted to use MonoAlg3D to run my simulations with my own mesh, but I have had problems with configuring the domain. When I run it, it gives me the error 'configuring the tissue domain'. I use Hexa-mesh-from-VTk to convert my vtu mesh to alg. Can you help me with the domain configuration??

**Comment by rsachetto** [comment]
Hi Benja,

Can you send me your mesh or the configuration file you are using to load the mesh?

**Reply by benjamitchell** [reply]
I think I was able to solve my problem, however, I still don't fully understand the parameters of [domain]. This is my configuration:
[main]
num_threads=1
dt_pde=0.02
simulation_time=500.0
abort_on_no_activity=false
use_adaptivity=false

[update_monodomain]
main_function=update_monodomain_default

[save_result]
print_rate=100
output_dir=tests/outputs
main_function=save_as_vtu
init_function=init_save_as_vtk_or_vtu
end_function=end_save_as_vtk_or_vtu
save_pvd=true
file_prefix=Test

[assembly_matrix]
init_function=set_initial_conditions_fvm
sigma_x=0.0001334
sigma_y=0.0001334
sigma_z=0.0001334
library_file=shared_libs/libdefault_matrix_assembly.so
main_function=homogeneous_sigma_assembly_matrix

[linear_system_solver]
tolerance=1e-16
use_preconditioner=no
use_gpu=yes
max_iterations=200
library_file=shared_libs/libdefault_linear_system_solver.so
main_function=conjugate_gradient
init_function=init_conjugate_gradient

[ode_solver]
dt=0.02
use_gpu=true
gpu_id=0
library_file=shared_libs/libten_tusscher_2006.so

[domain]
name=Custom Mesh
mesh_file=meshes/output_test2.alg
start_discretization = 1000.0
maximum_discretization = 500.0
number_of_points = 101896
num_extra_fields = 0
main_function=initialize_grid_with_custom_mesh

[stim_base]
z_limit=1000.0
start = 0.0
duration = 2.0
current = -50.0
main_function= stim_if_z_less_than

**Comment by benjamitchell** [comment]
Hi again Rafael! jajaja
I have some questions about hexa-mesh-from-VTK, in particular, how to include the fibers. Do you have any example or something that could help me?
I hope it's okay that I ask you about hexa-mesh here

**Reply by rsachetto** [reply]
No problem asking about the converter here. To include the fibers, you have to first extract them from the vtk file by configuring a .ini file. In this link https://github.com/rsachetto/MonoAlg3D_C/wiki/Loading-a-VTU-mesh we have some instructions on how to do it. 

On your config I can se:

start_discretization = 1000.0
maximum_discretization = 500.0

1000um discretization seems big. Maybe your start discretization is 500um?

**Reply by benjamitchell** [reply]
Thank you very much, Rafael, it helped me a lot

**Reply by benjamitchell** [reply]
Hi Rafael, it's me again! jajaja
I was able to create a .alg file with the fibers in hexa-mesh. It has this format:

center_x, center_y, center_z, side_length_x, side_length_y, side_length_z, tag_data, fiber_x, fiber_y, fiber_z

I can't load the fibers into the simulation, I followed the Load-A-VTU-Mesh procedure, but it still doesn't work. Could you help me? I'm sorry to bother you again.

**Reply by rsachetto** [reply]
Hi Benja, could you send me a sample of your mesh and the config file you are trying to use? You can send maybe the firt half or less of your mesh file, so I can try to debug the problem.

**Reply by benjamitchell** [reply]
Yes!! I sent you the files in the email!

**Reply by benjamitchell** [reply]
I write you again!
